nni/algorithms/compression/tensorflow/pruning/one_shot.py

```diff
@@ -57,15 +62,54 @@ def calc_masks(self, sparsity, wrapper, wrapper_idx=None):
 
             w_abs = tf.math.abs(weight)
             k = tf.size(weight) - num_prune
-            topk = tf.math.top_k(tf.reshape(w_abs, [-1]), k)[0]
+            topk = tf.math.top_k(tf.reshape(w_abs, [-1]), k).values
             if tf.size(topk) == 0:
                 mask = tf.zeros_like(weight)
             else:
                 mask = tf.math.greater_equal(w_abs, topk[-1])
             masks[weight_variable.name] = tf.cast(mask, weight.dtype)
         return masks
 
+class SlimPrunerMasker(WeightMasker):
+    def __init__(self, model, pruner, **kwargs):
+        super().__init__(model, pruner)
+        weight_list = []
+        for wrapper in pruner.wrappers:
+            weights = [w for w in wrapper.layer.weights if '/gamma:' in w.name]
+            assert len(weights) == 1, f'Bad weights: {[w.name for w in wrapper.layer.weights]}'
+            weight_list.append(tf.math.abs(weights[0].read_value()))
+        all_bn_weights = tf.concat(weight_list, 0)
+        k = int(all_bn_weights.shape[0] * (1 - pruner.wrappers[0].config['sparsity']))
```

Comment:
Why use `shape[0]` here which represents the first dimension of `all_bn_weights`? If we want to pick (1-sparsity) max values from weight, shouldn't we use count of all elements in `all_bn_weights` instead of `shape[0]`?

Reply:
The value k is correct if it represents preserved channel. But then the calculation of `top_k` and `self.global_threshold` are wrong.

Reply:
Current implementation is correct since this pruner only prune BatchNormalization layer....

Reply:
I agree it looks strange, but that's how PyTorch version works.
What's your suggestion?

Reply:
I think this is because SlimPruner is a channel level pruner, so by default, we use `shape[0]`. Maybe can refactor if we support more kind of channel scaling factor, but for now, it is good enough.

Reply:
I see, SlimPruner is for pruning BatchNorm operator whose number of weight parameter is equal to channel number. In this case, all weight parameters are in one dimension and pruning each of them equal to pruning convolution channel. So current implementation is actually enough.